checks/shell_download_validate.go

```diff
@@ -605,24 +618,64 @@ func recordFetchFileFromString(cmd string) (fmap map[string]bool, ok bool, err e
 	return files, true, err
 }
 
-func validateCommandIsNotFileExecute(cmd, pathfn string, files map[string]bool,
+func validateShellCommand(cmd, pathfn string, downloadedFiles map[string]bool,
 	logf func(s string, f ...interface{})) (bool, error) {
-	in := strings.NewReader(cmd)
-	f, err := syntax.NewParser().Parse(in, "")
+	ret := true
+
+	// First, check if the command launches an interpreter with
+	// a string, such as `bash -c "CMD"`, `python -c "CMD"`, `su -c "CMD"`, etc.
+	c, ok, err := extractInterpreterCommandFromString(cmd)
 	if err != nil {
-		return false, ErrParsingShellCommand
+		return false, err
+	} else if ok {
+		cmd = c
 	}
 
-	cmdValidated := true
-	syntax.Walk(f, func(node syntax.Node) bool {
-		// Check if we're calling a file we previously downloaded.
-		if isExecuteFiles(node, cmd, pathfn, files, logf) {
-			cmdValidated = false
-		}
+	// Validate it's not downloading and piping into a shell, like
+	// `curl | bash` (supports `sudo`).
+	r, err := validateCommandIsNotFetchPipeExecute(cmd, pathfn, logf)
+	if err != nil {
+		return false, err
+	} else if !r {
+		ret = false
+	}
+
+	// Validate it is not a download command followed by
+	// an execute: `curl > /tmp/file && /tmp/file`
+	//			   `curl > /tmp/file && bash /tmp/file`
+	//			   `curl > /tmp/file; bash /tmp/file`
+	//			   `curl > /tmp/file; /tmp/file`
+	// (supports `sudo`).
+	r, err = validateCommandIsNotFetchToFileExecute(cmd, pathfn, logf)
+	if err != nil {
+		return false, err
+	} else if !r {
+		ret = false
+	}
 
-		// Continue walking the node graph.
-		return true
-	})
+	// Validate it's not shelling out by redirecting input to stdin, like
+	// `bash <(wget -qO- http://website.com/my-script.sh)`. (supports `sudo`).
+	r, err = validateCommandIsNotFetchToStdinExecute(cmd, pathfn, logf)
+	if err != nil {
+		return false, err
+	} else if !r {
+		ret = false
+	}
 
-	return cmdValidated, nil
+	// TODO(laurent): add check for cat file | bash
+	// TODO(laurent): detect downloads of zip/tar files containing scripts.
```

Comment:
I think it's actually fine to stop here and not worry about these cases. Checking these accurately will be complex, making maintenance of this difficult and introducing more potential false positives.